consensus/istanbul/core/roundchange.go

```diff
@@ -48,45 +48,125 @@ func (c *core) sendRoundChange(round *big.Int) {
+		// Verify message signed by a validator
+		data, err := message.PayloadNoSig()
+		if err != nil {
+			return err
+		}
+
+		signer, err := c.validateFn(data, message.Signature)
+		if err != nil {
+			return err
+		}
+
+		if signer != message.Address {
+			return errInvalidRoundChangeCertificateMsgSignature
+		}
+
+		// Check for duplicate ROUND CHANGE messages
+		if seen[signer] {
+			return errInvalidRoundChangeCertificateDuplicate
+		}
+		seen[signer] = true
+
+		// Check that the message is a ROUND CHANGE message
+		if istanbul.MsgRoundChange != message.Code {
+			return errInvalidRoundChangeCertificateMsgCode
+		}
+
+		var roundChange *istanbul.RoundChange
+		if err := message.Decode(&roundChange); err != nil {
+			logger.Error("Failed to decode ROUND CHANGE in certificate", "err", err)
+			return err
+		}
+
+		// Verify ROUND CHANGE message is for a proper view
+		if err := c.checkMessage(istanbul.MsgRoundChange, roundChange.View); err != nil {
+			return errInvalidRoundChangeCertificateMsgView
+		}
+
+		decodedMessages[i] = *roundChange
+	}
+
+	for i, message := range roundChangeCertificate.RoundChangeMessages {
+		// No need to process ROUND CHANGE messages we've seen before.
+		if !c.backend.IsKnownMessage(message) {
+			err := c.handleDecodedCheckedRoundChange(&message, &decodedMessages[i])
+			// We want to continue to process ROUND CHANGE messages if they're for future rounds.
+			// TODO(asa): Should we just process every message regardless of if there's an error?
+			if err != nil && err != errIgnored {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (c *core) handleRoundChange(msg *istanbul.Message) error {
+	logger := c.logger.New("state", c.state, "from", msg.Address, "cur_round", c.current.Round(), "cur_seq", c.current.Sequence(), "func", "handleRoundChange", "tag", "handleMsg")
 
 	// Decode ROUND CHANGE message
-	var rc *istanbul.Subject
+	var rc *istanbul.RoundChange
 	if err := msg.Decode(&rc); err != nil {
 		logger.Error("Failed to decode ROUND CHANGE", "err", err)
 		return errInvalidMessage
 	}
 
-	if err := c.checkMessage(msgRoundChange, rc.View); err != nil {
+	if err := c.checkMessage(istanbul.MsgRoundChange, rc.View); err != nil {
+		logger.Info("Check round change message failed", "err", err)
 		return err
 	}
 
+	return c.handleDecodedCheckedRoundChange(msg, rc)
+}
+
+func (c *core) handleDecodedCheckedRoundChange(msg *istanbul.Message, rc *istanbul.RoundChange) error {
+	logger := c.logger.New("state", c.state, "from", msg.Address, "cur_round", c.current.Round(), "cur_seq", c.current.Sequence(), "func", "handleDecodedCheckedRoundChange", "tag", "handleMsg")
+
 	cv := c.currentView()
 	roundView := rc.View
 
+	// Handle the PREPARED certificate if present.
+	var num int
+	if rc.HasPreparedCertificate() {
+		if err := c.handlePreparedCertificate(rc.PreparedCertificate); err != nil {
+			// TODO(asa): Should we still accept the round change message without the certificate if this fails?
```

Comment:
Wdyt @trianglesphere? My hunch is no?

Reply:
I think it doesn't really matter given again **honest** nodes **should not** send invalid prepared certificates. I think that this could open up a minor risk of byzantine nodes trying to split honest node's rounds. I also think cutting the `prepared certificate` when invalid shouldn't hurt much.